README.md
App-GitHooks-Plugin-BlockNOCOMMIT
=================================

[![Build Status](https://travis-ci.org/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT.svg?branch=master)](https://travis-ci.org/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT)
[![Coverage Status](https://coveralls.io/repos/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT/badge.svg?branch=master)](https://coveralls.io/r/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT?branch=master)
[![CPAN](https://img.shields.io/cpan/v/App-GitHooks-Plugin-BlockNOCOMMIT.svg)](https://metacpan.org/release/App-GitHooks-Plugin-BlockNOCOMMIT)
[![License](https://img.shields.io/badge/license-GPLv3-blue.svg)](https://opensource.org/licenses/GPL-3.0)

App::GitHooks plugin to prevent committing code with #NOCOMMIT mentions.


INSTALLATION
------------

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install


SUPPORT AND DOCUMENTATION
-------------------------

After installing, you can find documentation for this module with the
perldoc command.

	perldoc App::GitHooks::Plugin::BlockNOCOMMIT


You can also look for information at:

 * [GitHub's request tracker (report bugs here)]
   (https://github.com/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT/issues)

 * [AnnoCPAN, Annotated CPAN documentation]
   (http://annocpan.org/dist/App-GitHooks-Plugin-BlockNOCOMMIT)

 * [CPAN Ratings]
   (http://cpanratings.perl.org/d/App-GitHooks-Plugin-BlockNOCOMMIT)

 * [MetaCPAN]
   (https://metacpan.org/release/App-GitHooks-Plugin-BlockNOCOMMIT)


LICENSE AND COPYRIGHT
---------------------

Copyright (C) 2013-2016 Guillaume Aubert

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License version 3 as published by the Free
Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see http://www.gnu.org/licenses/



Changes
Revision history for App-GitHooks-Plugin-BlockNOCOMMIT

v1.1.0  2016-05-21
        - Switched to Test::Requires::Git instead of the now obsolete
          Test::Git::has_git(). Thank you, Philippe Bruhat!
        - Relicensed under Perl 5 terms.
        - Cleaned up dependencies.

v1.0.2  2014-06-07
        - Made git test safer to run.

v1.0.1  2014-06-04
        - Noted git version in tests to help troubleshooting.
        - Required git v1.7.4.1.

v1.0.0  2014-05-31
        - Public release.


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.010;
use ExtUtils::MakeMaker;
WriteMakefile
(
  'test' => {
              'TESTS' => 't/*.t'
            },
  'EXE_FILES' => [],
  'NAME' => 'App::GitHooks::Plugin::BlockNOCOMMIT',
  'INSTALLDIRS' => 'site',
  'PL_FILES' => {},
  'PREREQ_PM' => {
                   'Data::Validate::Type' => 0,
                   'Test::More' => '0.94',
                   'Test::FailWarnings' => 0,
                   'App::GitHooks' => 0,
                   'Git::Repository' => 0,
                   'Test::Type' => 0,
                   'Carp' => 0,
                   'File::Slurp' => 0,
                   'Capture::Tiny' => 0,
                   'Test::Requires::Git' => '1.005',
                   'Test::Exception' => 0
                 },
  'VERSION_FROM' => 'lib/App/GitHooks/Plugin/BlockNOCOMMIT.pm'
)
;


lib/App/GitHooks/Plugin/BlockNOCOMMIT.pm
package App::GitHooks::Plugin::BlockNOCOMMIT;

use strict;
use warnings;

use base 'App::GitHooks::Plugin';


# External dependencies.
use Carp;
use Data::Validate::Type;
use File::Slurp ();

# Internal dependencies.
use App::GitHooks::Constants qw( :PLUGIN_RETURN_CODES );


=head1 NAME

App::GitHooks::Plugin::BlockNOCOMMIT - Prevent committing code with #NOCOMMIT mentions.


=head1 DESCRIPTION

Sometimes you want to experiment with code, but you want to make sure that test
code doesn't get committed by accident.

This plugin allows you to use C<#NOCOMMIT> to indicate such code blocks, and
will prevent you from committing those blocks.

For example:

	# This is a test that will not work once deployed.
	# NOCOMMIT
	...

Note that the following variations on the tag are correctly picked up:

=over 4

=item * C<#NOCOMMIT>

=item * C<# NOCOMMIT>

=item * C<# NO COMMIT>

=back


=head1 VERSION

Version 1.1.0

=cut

our $VERSION = '1.1.0';


=head1 METHODS

=head2 get_file_pattern()

Return a pattern to filter the files this plugin should analyze.

	my $file_pattern = App::GitHooks::Plugin::BlockNOCOMMIT->get_file_pattern(
		app => $app,
	);

=cut

sub get_file_pattern
{
	# Check all files.
	return qr//;
}


=head2 get_file_check_description()

Return a description of the check performed on files by the plugin and that
will be displayed to the user, if applicable, along with an indication of the
success or failure of the plugin.

	my $description = App::GitHooks::Plugin::BlockNOCOMMIT->get_file_check_description();

=cut

sub get_file_check_description
{
	return 'The file has no #NOCOMMIT tags.';
}


=head2 run_pre_commit_file()

Code to execute for each file as part of the pre-commit hook.

	my $success = App::GitHooks::Plugin::BlockNOCOMMIT->run_pre_commit_file();

=cut

sub run_pre_commit_file
{
	my ( $class, %args ) = @_;
	my $app = delete( $args{'app'} );
	my $file = delete( $args{'file'} );
	my $git_action = delete( $args{'git_action'} );
	croak 'Unknown argument(s): ' . join( ', ', keys %args )
		if scalar( keys %args ) != 0;

	# Check parameters.
	croak "The 'app' argument is mandatory"
		if !Data::Validate::Type::is_instance( $app, class => 'App::GitHooks' );
	croak "The 'file' argument is mandatory"
		if !defined( $file );
	croak "The 'git_action' argument is mandatory"
		if !defined( $git_action );

	# Ignore deleted files.
	return $PLUGIN_RETURN_SKIPPED
		if $git_action eq 'D';

	# Ignore merges, since they correspond mostly to code written by other people.
	my $staged_changes = $app->get_staged_changes();
	return $PLUGIN_RETURN_SKIPPED
		if $staged_changes->is_merge();

	# Ignore revert commits.
	return $PLUGIN_RETURN_SKIPPED
			if $staged_changes->is_revert();

	# Determine what lines were written by the current user.
	my @review_lines = ();
	if ( $git_action eq 'A' )
	{
		# "git blame" fails on new files, so we need to add the entire file
		# separately.
		my @lines = File::Slurp::read_file( $file );
		foreach my $i ( 0 .. scalar( @lines ) - 1 ) {
			chomp( $lines[$i] );
			push(
				@review_lines,
				{
					line_number => $i,
					code        => $lines[$i],
				}
			);
		}
	}
	else
	{
		# Find uncommitted lines only.
		my $repository = $app->get_repository();
		my $blame_lines = $repository->blame(
			$file,
			use_cache => 1,
		);

		foreach my $blame_line ( @$blame_lines )
		{
			my $commit_attributes = $blame_line->get_commit_attributes();
			next unless $commit_attributes->{'author-mail'} eq 'not.committed.yet';
			push(
				@review_lines,
				{
					line_number => $blame_line->get_line_number(),
					code        => $blame_line->get_line(),
				}
			);
		}
	}

	# Inspect uncommitted lines for NOCOMMIT mentions.
	my @violations = ();
	foreach my $line ( @review_lines )
	{
		my $code = $line->{'code'};
		next if $code !~ /
			(?:\/\/|\#)      # Support Perl and Javascript comments.
			.*               # Support NOCOMMIT anywhere in the comment, not just at the beginning.
			\bNO\s*COMMIT\b  # Support NOCOMMIT and NO COMMIT.
		/ix;

		my $bad_line = $code;
		$bad_line =~ s/\b(NO\s*COMMIT)\b/$app->color('red', $1)/iegs;
		push(
			@violations,
			sprintf( "line #%d:   %s\n", $line->{'line_number'}, $bad_line )
		);
	}

	die "#NOCOMMIT tags found:\n" . join( '', @violations ) . "\n"
			if scalar( @violations ) != 0;

	return $PLUGIN_RETURN_PASSED;
}


=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/guillaumeaubert/App-GitHooks-Plugin-BlockNoCommit/issues/new>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

	perldoc App::GitHooks::Plugin::BlockNOCOMMIT


You can also look for information at:

=over

=item * GitHub's request tracker

L<https://github.com/guillaumeaubert/App-GitHooks-Plugin-BlockNoCommit/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/app-githooks-plugin-blocknocommit>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/app-githooks-plugin-blocknocommit>

=item * MetaCPAN

L<https://metacpan.org/release/App-GitHooks-Plugin-BlockNoCommit>

=back


=head1 AUTHOR

L<Guillaume Aubert|https://metacpan.org/author/AUBERTG>,
C<< <aubertg at cpan.org> >>.


=head1 COPYRIGHT & LICENSE

Copyright 2013-2016 Guillaume Aubert.

This code is free software; you can redistribute it and/or modify it under the
same terms as Perl 5 itself.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the LICENSE file for more details.

=cut

1;


Build.PL
use 5.006;
use strict;
use warnings;

use Module::Build;


my $class = Module::Build->subclass(
	code =>
	q|
		# Override 'test' to support running extended tests in the xt/
		# directory.
		sub ACTION_test
		{
			my ( $self ) = @_;

			$self->recursive_test_files(1);

			$self->test_files( 't', 'xt' )
				if $ENV{'RELEASE_TESTING'};

			return $self->SUPER::ACTION_test();
		}

		# Force running extended tests when testing the distribution.
		sub ACTION_disttest
		{
			my ( $self ) = @_;

			local $ENV{ RELEASE_TESTING } = 1;

			return $self->SUPER::ACTION_disttest();
		}
	|,
);

my $builder = $class->new(
	module_name          => 'App::GitHooks::Plugin::BlockNOCOMMIT',
	license              => 'GPL_3',
	dist_author          => q{Guillaume Aubert <aubertg@cpan.org>},
	dist_version_from    => 'lib/App/GitHooks/Plugin/BlockNOCOMMIT.pm',
	build_requires       =>
	{
		'Capture::Tiny'        => 0,
		'Git::Repository'      => 0,
		'Test::Exception'      => 0,
		'Test::FailWarnings'   => 0,
		'Test::Requires::Git'  => 1.005,
		'Test::More'           => 0.94,
		'Test::Type'           => 0,
	},
	requires             =>
	{
		'App::GitHooks'        => 0,
		'Carp'                 => 0,
		'File::Slurp'          => 0,
		'Data::Validate::Type' => 0,
		'perl'                 => 5.010,
	},
	add_to_cleanup       =>
	[
		'App-GitHooks-Plugin-BlockNOCOMMIT-*',
	],
	create_makefile_pl   => 'traditional',
	configure_requires   =>
	{
		'Module::Build'        => 0,
	},
	meta_merge           =>
	{
		resources              =>
		{
			repository  => 'https://github.com/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT',
			homepage    => 'https://metacpan.org/release/App-GitHooks-Plugin-BlockNOCOMMIT',
			bugtracker  => 'https://github.com/guillaumeaubert/App-GitHooks-Plugin-BlockNOCOMMIT/issues',
		},
	},
	recursive_test_files => 1,
);

$builder->create_build_script();
